File: Programmers/C30L12983/C30L12983.cpp
#include <algorithm>
#include <iostream>
#include <map>
#include <string>
#include <vector>
using namespace std;
int dp_dfs(vector<string> &table, map<ptrdiff_t, int> &dpc,
           string::const_iterator be, string::const_iterator ed) {
    if (be == ed) {
        return 0;
    }
    // string t(be, ed);
    if (dpc.find(distance(be, ed)) != dpc.cend()) {
        return dpc[distance(be, ed)];
    }
    int result = -1;
    for (auto &w : table) {
        if (equal(w.begin(), w.end(), be)) {
            int tmp = dp_dfs(table, dpc, be + w.size(), ed);
            if (tmp != -1) {
                if (result == -1) {
                    result = tmp + 1;
                } else {
                    result = min(result, tmp + 1);
                }
            }
        }
    }
    dpc[distance(be, ed)] = result;
    return result;
}

int solution(vector<string> strs, string t) {
    map<ptrdiff_t, int> caches;
    sort(strs.begin(), strs.end(),
         [](auto a, auto b) { return a.size() > b.size(); });
    int answer = dp_dfs(strs, caches, t.cbegin(), t.cend());
    return answer;
}
int main(int argc, char const *argv[]) {
    // cout << solution({"ba", "an", "nan", "ban", "n"}, "banana") << endl;
    cout << solution({"app", "ap", "p", "l", "e", "ple", "pp"}, "apple")
         << endl;
    return 0;
}

File: Programmers/C30L12983/C30L12983.py
# %%
import functools as ft
import itertools as it
import math

# BFS
# def solution(strs, t):
#     words = [(w, 1) for w in strs]
#     q = [(0, t)]
#     while q:
#         d, tmp = q.pop(0)
#         if len(tmp) == 0:
#             return d
#         for w, v in words:
#             if tmp.startswith(w):
#                 q.append((d + v, tmp[len(w):]))
#     return -1

# 
# 어캐풀엇누
def solution(strs, t):
    n = len(t)
    memo = [float("inf")] * (n + 1)
    memo[0] = 0
    sizes = set(len(s) for s in strs)
    strs = set(strs)
    for i in range(n + 1):
        for size in sizes:
            if i + size < n + 1 and t[i:i + size] in strs:
                memo[i + size] = min(memo[i + size], memo[i] + 1)
    return memo[n] if memo[n] < float("inf") else -1

print(solution(["ba", "na", "n", "a"], "banana"))
# print(solution(["ba", "na", "n", "a"], "ana"))
# print(solution(["app", "ap", "p", "l", "e", "ple", "pp"], "apple"))
# print(solution(["ba", "an", "nan", "ban", "n"], "banana"))
# print(solution(["ba", "an", "nan", "ban", "n"], "xv"))
